src/Mod/Sketcher/App/SketchObject.cpp

```diff
@@ -7574,27 +7574,61 @@ void SketchObject::rebuildExternalGeometry()
                 BRepAdaptor_Curve curve(edge);
                 if (curve.GetType() == GeomAbs_Line) {
                     ExternalGeo.push_back(projectLine(curve, gPlane, invPlm));
+                    break;
+                }
+
+                gp_Dir vec1 = sketchPlane.Axis().Direction();
+                gp_Dir vec2;
+                if (curve.GetType() == GeomAbs_Circle) {
+                    vec2 = curve.Circle().Axis().Direction();
+                }
+                else if (curve.GetType() == GeomAbs_Ellipse) {
+                    vec2 = curve.Ellipse().Axis().Direction();
                 }
-                else if (curve.GetType() == GeomAbs_Circle) {
-                    gp_Dir vec1 = sketchPlane.Axis().Direction();
-                    gp_Dir vec2 = curve.Circle().Axis().Direction();
+
+                gp_Pnt beg, end;
+                if ((curve.GetType() == GeomAbs_Circle || curve.GetType() == GeomAbs_Ellipse)
+                    && !curve.IsClosed()) {
+                    beg = ProjPointOnPlane_XYZ(curve.Value(curve.FirstParameter()), sketchPlane);
+                    end = ProjPointOnPlane_XYZ(curve.Value(curve.LastParameter()), sketchPlane);
+                    Base::Vector3d vBeg(beg.X(), beg.Y(), beg.Z());
+                    Base::Vector3d vEnd(end.X(), end.Y(), end.Z());
+                    invPlm.multVec(vBeg, vBeg);
+                    invPlm.multVec(vEnd, vEnd);
+                    beg = gp_Pnt(vBeg[0], vBeg[1], vBeg[2]);
+                    end = gp_Pnt(vEnd[0], vEnd[1], vEnd[2]);
+
+                    if (vec1.IsNormal(vec2, Precision::Angular())) {
+                        if (beg.SquareDistance(end) < Precision::SquareConfusion()) {
+                            Base::Vector3d p = (vBeg + vEnd) / 2;
+                            Part::GeomPoint* point = new Part::GeomPoint(p);
+                            ExternalGeo.push_back(point);
+                        }
+                        else {
+                            Part::GeomLineSegment* line = new Part::GeomLineSegment();
+                            line->setPoints(vBeg, vEnd);
+                            GeometryFacade::setConstruction(line, true);
+                            ExternalGeo.push_back(line);
+                        }
+                        break;
+                    }
+                }
+
+                if (curve.GetType() == GeomAbs_Circle) {
                     if (vec1.IsParallel(vec2, Precision::Confusion())) {

```

Comment:
Do you think this should be `Precision::Angular()` instead?

Reply:
Interesting thing to bring up. In theory since the tolerance argument ends up being used to compare angles in radians then that suggests using `Precision::Angular()` instead.

In practice I think that may introduce problems. This code was introduced in https://github.com/FreeCAD/FreeCAD/commit/fc903d93dfd545a0a11be9187f8641f6e719a2d9 which was 10 years ago. I checked and `Precision::Angular()` existed at the time but they didn't use it (can't tell if intentional). Over 10 years, that's many existing files that could break if they contained a circle that now becomes an ellipse.

I checked LinkStable branch and there's actually a mention of an issue with using 'Precision::AngleConfusion()' (by which I assume they are referring to `Precision::Angular()`):
https://github.com/realthunder/FreeCAD/blob/LinkStable/src/Mod/Sketcher/App/SketchObject.cpp#L8526
Namely if the angle is very small but still bigger than the tolerance then the projection will still be a circle instead of an ellipse.

So given this is existing code and I don't want to potentially introduce issues I'd prefer to not change it.

Reply:
I think we should be fine. The same arguments can be made for the `SquareConfusion` parameter used otherwise.

Additionally, I think it would be better that projections of circles are allowed to be ellipses, but not vice versa.

Reply:
It does not work to constrain a circle equal to an ellipse. So circle->ellipse can break files.

> The same arguments can be made for the SquareConfusion parameter used otherwise.

No. Sketches do change but they shouldn't generally break. Checking against `SquareConfusion` is stricter than against `Confusion` and given how it's used in the code the switch results in less degenerate geometry, i.e. more geometry. For example it's fine to change circles into arcs as the 2 extra points don't break anything.

Reply:
> It does not work to constrain a circle equal to an ellipse. So circle->ellipse can break files.

I expect this breakage can also happen by changing some other variables (for example if there were a datum plane made at an angle, but only coincidentally parallel to the projected circle). In general, such constraints with external edges should be discouraged in the first place.

Regardless, I should note that this isn't a blocker for this PR, and can be handled later.